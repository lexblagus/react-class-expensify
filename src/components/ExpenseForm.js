import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



//console.log(  moment().format('MMM Do')  );

export default class ExpenseForm extends React.Component{
	constructor(props){
		super(props);
		this.state = {
			id: props.expense ? props.expense.id : undefined,
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? (props.expense.amount/100).toString() : 0,
			notes: props.expense ? props.expense.notes : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: ''
		};
	}
	onDescChange = (evt) => {
		const description = evt.target.value;
		this.setState(() => ({description}));
	};
	onAmountChange = (evt) => {
		const amount = evt.target.value;
		if(
			!amount ||
			amount.match(/^\d{1,}(\.\d{0,2})?$/)
		){
			//console.log('VALID');
			this.setState(() => ({amount}));
		}else{
			//console.log('INvalid');
		}
	};
	onDateChange = (createdAt) => {
		if( createdAt ){
			this.setState(
				() => ({createdAt})
			);
		}
	};
	onNotesChange = (evt) => {
		const notes = evt.target.value;
		this.setState(() => ({notes}));
	};
	onFocusChange = ({ focused }) => {
		this.setState(
			() => ({ calendarFocused: focused })
		)
	};
	onSubmit = (e) => {
		e.preventDefault();
		
		if(
			!this.state.description ||
			!this.state.amount
		){
			this.setState(
				() => ({ error: 'missing stuff' })
			)
		}else{
			this.setState(
				() => ({ error: '' })
			)
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				notes: this.state.notes
			});
		}
	};
	render(){
		return (
			<div>
				{
					this.state.error &&
					<code>{this.state.error}</code>
				}
				<form onSubmit={this.onSubmit}>
					description: 
					<input
						type="text"
						placeholder="description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescChange}
					/>
					<br />
					notes: 
					<textarea
						placeholder="notes"
						value={this.state.notes}
						onChange={this.onNotesChange}
					/>
					<br />
					created at: 
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={()=>false}
					/>
					<br />
					amount: 
					<input
						type="number"
						placeholder="amount"
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<br />
					<button>Save</button>
				</form>
			</div>
		);
	}
	
}