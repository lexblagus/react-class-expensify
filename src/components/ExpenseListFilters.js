import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters';


const ExpenseListFilters = (props) => (
	<div>
		<input
			type="text"
			value={props.filters.text}
			onChange={(e) => {
				props.dispatch(setTextFilter(e.target.value));
			}}
			
		/>
		<select
			value={props.filters.sortBy}
			onChange={(evt) => {
				if(evt.target.value === 'date'){
					props.dispatch(sortByDate());
				} else if(evt.target.value === 'amount'){
					props.dispatch(sortByAmount());
				}
			}}
		>
			<option value="amount">Amount</option>
			<option value="date">Date</option>
		</select>
	</div>
);

const mapStateToProps = (state) => ({
	filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);
