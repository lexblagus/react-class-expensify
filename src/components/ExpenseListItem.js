import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';


const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
	<div>
		<h2>{description}</h2>
		<button
			onClick = {() => {
				dispatch(removeExpense({id}));
			}}
		>remove</button>
		&nbsp;
		amount = {amount},
		createdAt = {createdAt}
	</div>
);

export default connect()(ExpenseListItem);
