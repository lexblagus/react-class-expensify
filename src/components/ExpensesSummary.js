import React from 'react';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';
import { connect } from 'react-redux';

export const ExpensesSummary = ({qtt,tot}) => (
	<div>
		<h3 className="title is-3">
			List
			<span className="title is-4">
				{' '}
				with
				{' '}
				{qtt}
				{' '}
				item{qtt>1?'s':''}
				{' '}
				$ {tot}
			</span>
		</h3>
	</div>
);

const mapStateToProps = (state) => ({
		'tot': expensesTotal(selectExpenses(state.expenses,state.filters)),
		'qtt': selectExpenses(state.expenses,state.filters).length
});


export default connect(mapStateToProps)(ExpensesSummary);

//export default ExpensesSummary;