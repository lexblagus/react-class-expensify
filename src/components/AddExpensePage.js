import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';


const AddExpensePage = (props) => (
	<div>
		<div className="hero is-light">
			<h1 className="title is-1">Add expense</h1>
			<ExpenseForm
				onSubmit = {
					(expense)=>{
						props.dispatch(
							addExpense(expense)
						);
						props.history.push('/');
					}
				}
			/>
		</div>
	</div>
);

export default connect()(AddExpensePage);
