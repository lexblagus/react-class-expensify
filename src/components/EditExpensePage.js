import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
	return (
		<div>
			<section className="section">
				<div className="container">
					{
						props.expense &&
						<div>
							<h2 className="title is-2">
								Edit
								<span className="title is-4">
									&nbsp;
									{props.expense.id}
								</span>
							</h2>
							<div className="card">
								<div className="card-content">
									<ExpenseForm
										expense={props.expense}
										onSubmit={
											(expense)=>{
												console.log('updated expense',expense);
												props.dispatch(
													editExpense(
														props.expense.id,
														{...expense}
													)
												);
												props.history.push('/');
											}
										}
									/>
									<button
										className="button is-small is-danger"
										onClick = {() => {
											props.dispatch(
												removeExpense(
													{id: props.expense.id}
												)
											);
											props.history.push('/');
										}}
									>Delete</button>
								</div>
							</div>
						</div>
					}
					{
						!props.expense &&
						<div>
							<h2 className="title is-2">
								Item not found
							</h2>
						</div>
					}
				</div>
			</section>
		</div>
	);
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find(
		(expense) => (expense.id === props.match.params.id)
	)
});

export default connect( mapStateToProps )( EditExpensePage );
