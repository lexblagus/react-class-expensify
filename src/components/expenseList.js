import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
	<div className="container">
		<h3 className="title is-3">
			List
			<span className="title is-4">
				{' '}
				with
				{' '}
				{props.expenses.length}
				{' '}
				item{props.expenses.length>1?'s':''}
			</span>
		</h3>
		{
			props.expenses.length === 0 ? (
				<div className="notification is-warning">
					nothing right now
				</div>
			) : (
				props.expenses.map(
					(expense) => (
						<ExpenseListItem key={expense.id} {...expense} />
					)
				)
			)
		}
	</div>
);

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses,state.filters)
});


export default connect(mapStateToProps)(ExpenseList);
