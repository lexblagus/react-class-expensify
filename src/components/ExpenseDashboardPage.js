import React from 'react';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import ExpenseList from './ExpenseList';


const ExpenseDashboardPage = () => (
	<div>
		<section className="section">
			<div className="container">
				<h2 className="title is-2">View</h2>
				<div className="card">
					<div className="card-content">
						<ExpenseListFilters />
						<ExpensesSummary />
						<ExpenseList />
					</div>
				</div>
			</div>
		</section>
	</div>
);

export default ExpenseDashboardPage;
