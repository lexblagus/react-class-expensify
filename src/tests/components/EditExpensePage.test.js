import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import storeDummyData from '../fixtures/storeDummyData.js';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
	editExpense = jest.fn(); // spie
	removeExpense = jest.fn(); // spie
	history = { push: jest.fn() };
	wrapper = shallow(
		<EditExpensePage
			editExpense={editExpense}
			removeExpense={removeExpense}
			history={history}
			expense={storeDummyData[4]}
		/>
	);
});

test('should render EditExpensePage correctly', ()=>{
	expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', ()=>{
	wrapper.find('ExpenseForm').prop('onSubmit')( storeDummyData[4] );
	expect( history.push ).toHaveBeenLastCalledWith('/');
	expect(editExpense).toHaveBeenLastCalledWith( storeDummyData[4].id, storeDummyData[4] );
});

test('should handle removeExpense', ()=>{
	wrapper.find('button').prop('onClick')( storeDummyData[4].id );
	expect( history.push ).toHaveBeenCalledWith('/');
});
