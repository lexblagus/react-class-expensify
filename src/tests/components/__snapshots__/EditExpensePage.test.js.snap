// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <section
    className="section"
  >
    <div
      className="container"
    >
      <div>
        <h2
          className="title is-2"
        >
          Edit
          <span
            className="title is-4"
          >
            Â 
            adb329fa-d15a-4605-8e09-44f544a69230
          </span>
        </h2>
        <div
          className="card"
        >
          <div
            className="card-content"
          >
            <ExpenseForm
              expense={
                Object {
                  "amount": 200,
                  "createdAt": 1597615000,
                  "description": "Description B",
                  "id": "adb329fa-d15a-4605-8e09-44f544a69230",
                  "notes": "notes notes notes 2",
                }
              }
              onSubmit={[Function]}
            />
            <button
              className="button is-small is-danger"
              onClick={[Function]}
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
